File: frontend/src/lib/api/workspaces.ts
import { apiClient } from "./client";

export interface Workspace {
  id: string;
  name: string;
  description?: string | null;
  color: string;
  paper_count: number;
  owner_id?: string;
  created_at: string;
  updated_at?: string;
}

export interface WorkspaceListResponse {
  items: Workspace[];
  total: number;
}

export interface WorkspaceActivity {
  documents: number;
  flashcards: number;
  quizzes: number;
  chats: number;
}

import { DocumentListResponse } from "./documents";

export const workspacesApi = {
  list: () => apiClient<WorkspaceListResponse>("/workspaces", {
    method: "GET",
  }),

  create: (data: { name: string; description?: string; color?: string }) =>
    apiClient<{ workspace: Workspace }>("/workspaces", {
      method: "POST",
      body: JSON.stringify(data),
    }),

  get: (id: string) =>
    apiClient<Workspace>(`/workspaces/${id}`, {
      method: "GET",
    }),

  update: (id: string, data: { name?: string; description?: string; color?: string }) =>
    apiClient<Workspace>(`/workspaces/${id}`, {
      method: "PUT",
      body: JSON.stringify(data),
    }),

  delete: (id: string) =>
    apiClient<{ status: string }>(`/workspaces/${id}`, {
      method: "DELETE",
    }),

  getDocuments: (id: string) =>
    apiClient<DocumentListResponse>(`/workspaces/${id}/documents`, {
      method: "GET",
    }),

  getActivity: (id: string) =>
    apiClient<WorkspaceActivity>(`/workspaces/${id}/activity`, {
      method: "GET",
    }),
};

File: backend/app/schemas/workspaces.py
"""Workspace schemas."""

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


class WorkspaceBase(BaseModel):
    """Base workspace schema."""

    name: str = Field(..., max_length=128)
    description: str | None = Field(None, max_length=1000)
    color: str = Field(default="#6366f1", max_length=32)


class WorkspaceCreate(WorkspaceBase):
    """Schema for creating a workspace."""
    pass


class WorkspaceUpdate(BaseModel):
    """Schema for updating a workspace."""

    name: str | None = Field(None, max_length=128)
    description: str | None = Field(None, max_length=1000)
    color: str | None = Field(None, max_length=32)


class WorkspaceResponse(WorkspaceBase):
    """Schema for workspace response."""

    id: UUID
    owner_id: UUID
    paper_count: int
    created_at: datetime
    updated_at: datetime | None = None

    model_config = ConfigDict(from_attributes=True)


class WorkspaceListResponse(BaseModel):
    """Schema for listing workspaces."""

    items: list[WorkspaceResponse]


class WorkspaceActivityResponse(BaseModel):
    """Schema for workspace activity metrics."""

    documents: int
    flashcards: int
    quizzes: int
    chats: int
